### PR TITLE
changefeedccl: Rework test feed implementation.

### PR DESCRIPTION
**Pull request by miretskiy**
Rework TestFeed implementation used in changfeed tests.
The goal of this rework was to make TestFeeds more explicit
in how feeds are setup and managed.

In particular:
  1. Move TestFeed implementation into changfeedccl package.
     While having TestFeed live in cdctest was useful, the reality
     is that the implementations needed substantial amount of orchestration
     performed inside changefeedccl package.  As a result, it was
     very hard to understand interplay of various knobs/synchrnonization
     mechanisms setup in the ccl package with the code inside TestFeed package.
  2. Introduce EnterpriseTestFeed interface which is used when running enterprise
     feeds.  Simplify job error handling code.  Instead of polling for job error
     on every call to `Next()`, job feeds now arrange for a special `jobs.Resumer`
     to be used which reports errors directly, without relying on "flush knobs".
     In addition, all enterprise feeds now behave similarly wrt to job error
     handling as well as handling of multiple concurrently executing feeds.
  3. Instead of relying on testing knobs to synchronize Flush() events
     (`AfterSinkFlush`), inject a special Sink implementation that does synchronizations
     directly.
  4. Rework helpers used to start various test feed implementations to use shared
     configuration logic.
  5. Various dependency injections knobs are now scoped to each feed
     (as opposed to server level testing knobs).
  6. Enable additional tests for kafka/cloud storage.

Release Notes: None

### CODE REVIEW

**Comment by miretskiy** [comment]
@stevendanna ready for another look.

**Review by ajwerner — COMMENTED** [review]
First, high-level, certain to be annoying comment. How would you feel about putting `testfeed_test.go` into the `changefeedccl_test` package so it cannot use any unexported symbols and then for symbols you want it to use, create a `helpers_test.go` file that exports those symbols for its use? The cost of just pulling all of this code into the package is that it muddles the boundaries. I've come to really like boundaries. It's not clear to me from reading through the commit what really gets simplified here. If you at least broke it out into two commits. One that moves the code but leaves it only using exported symbols and then a second that cleans up the interfaces, the intent would be clearer. I think I still prefer the above suggestion but this would aid reviewing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-:-M_MRARV1nldiZn5kMxD:b-8ks3cx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

**Comment by miretskiy** [comment]
> How would you feel about putting testfeed_test.go into the changefeedccl_test package so it cannot use any unexported symbols and then for symbols you want it to use, create a helpers_test.go file that exports those symbols for its use?

I'm not sure this would be any better than keeping in the current `cdctest`.

> The cost of just pulling all of this code into the package is that it muddles the boundaries. I've come to really like boundaries. It's not clear to me from reading through the commit what really gets simplified here. 

I agree that boundaries are good.  And I have tried very hard to keep this boundary.  However, what I found was that
the tests we had were not actually "black box" tests -- but instead, the tests had kinda of a split personality, where
the cdctest package `TestFeed` had no chance of working without extensive orchestration and setup performed
inside `changefeedccl` package.  So, you have tests that already cross all sort of boundaries, but this criss crossing
is very subtle, and hard to follow.

In addition, for better or worse, I've added ability to have "kafka feed" -- that was done through more injection in the
`changefeedccl` package, and quite a bit of code to get the orchestration working correctly, while the rest of "kafka feed" was in the `cdctest` package. 

I feel that the we were breaking the boundaries in a sense -- and that having 1/2 the test code in the cdctest while the rest was in the helpers_test was not helpful -- if anything it hid the tight coupling between helpers_test and feed implementation.


I think it might be possible that have those boundaries -- but to do that I would also need to move few pieces into shared package (changefeedbase probably): at the very least testing_knobs and Sink interface.  I would also probably need to do something around "reportErrorResumer" so that it has access to OnPauseRequested.  I guess, all of this is underscoring just how intertwined the code is.  If you have very strong preference, I could try to undo some parts of this refactoring (and move required files/interfaces into shared packaged).  Would love to discuss this further with you...

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64693)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-M_QZG4w4UXYkE5bRQXa:-M_QZG4x1kSeE5ekNMyD:b1ue57l) (no related file):*
I created a branch with a change that moves the testfeed implementations and then applied everything else if anyone finds that easier to review.  (warning, that branch builds but I didn't test it further and I resolved the conflicts somewhat quickly so perhaps there are mistakes)

https://github.com/cockroachdb/cockroach/compare/master...stevendanna:ssd/miretskiy/test_feed

---

**Review by stevendanna — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64693)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-M_QawOkCbL4MV9KKbGD:-M_QawOkCbL4MV9KKbGE:b-pcxm27) (no related file):*
Thanks for walking me through some of this the other day.  From that conversation and what I can see, the biggest changes are (paraphrasing some of your commit message)

1) How we we wait for flush events: Rather than a single flushCh that is shared by all changefeeds (because it was written to in a server-level testing knob) we now have per-changefeed flush notifications via a knew "flush notifying sink".  Since sinks are created at multiple places when creating a changefeed, we have a new testing knob that intercepts the creation of sinks on a per job basis and allows us to wrap them with our flush-notifying implementation. 

2) How job errors are handled: Rather than polling the jobs table, we now have a Resumer implementation that wraps the main changefeed resumer and calls back into caller-provided code when the changefeed fails.

3) Since we want both of the above to happen on per-changefeed basis, we now have a `depInjector` which coordinates the above on a per-job basis.

I agree with Andrew that it might have been good to break this up into multiple commits to make it easier to review. But I also get how one change leads to the next when working in this code.

Since we are all working in the same relatively small pond at the moment, I'd err on the side of iterating on this on master.

Apologies, but while I spend a good amount of time reading this, I haven't left many in-code comments, since most of the time I found myself commenting on something it was existing code. I've marked it as approved since I'm not opposed to this direction and since it facilitated turning on more Kafka tests.

---
___
*[pkg/ccl/changefeedccl/testfeed_test.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-M_QdDho9bHmiAcYUjtj:-M_QdDho9bHmiAcYUjtk:b-5h49ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/aeca9fabddd06cd0cc041f315e50c09801c79ae5/pkg/ccl/changefeedccl/testfeed_test.go#L211)):*
> ```Go
> // OnFailOrCancel implements jobs.Resumer
> func (r *reportErrorResumer) OnFailOrCancel(ctx context.Context, execCtx interface{}) error {
> 	defer r.jobFailed()
> ```

It would be kinda nice if we got access to the error via OnFailOrCancel, then we could provide it as an argument to this callback. Anyway, that is just idle musing, not something to change.
___
*[pkg/ccl/changefeedccl/testfeed_test.go, line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-M_QaYF54kp4Qy46wWEi:-M_QaYF54kp4Qy46wWEj:bv75lwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/aeca9fabddd06cd0cc041f315e50c09801c79ae5/pkg/ccl/changefeedccl/testfeed_test.go#L368)):*
> ```Go
> // eventReady returns a channel that can be waited on until the next
> // event.
> func (s *sinkSynchronizer) eventReady() chan struct{} {
> ```

I could imagine this handling other events, but I could imagine naming this something like `nextFlush()` to make it clear at the call site that it waits for flushes specifically right now.

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64693)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-M_QawOkCbL4MV9KKbGD:-M_Qv3E31BNsuhvLicLz:b-teijr1) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Thanks for walking me through some of this the other day.  From that conversation and what I can see, the biggest changes are (paraphrasing some of your commit message)

1) How we we wait for flush events: Rather than a single flushCh that is shared by all changefeeds (because it was written to in a server-level testing knob) we now have per-changefeed flush notifications via a knew "flush notifying sink".  Since sinks are created at multiple places when creating a changefeed, we have a new testing knob that intercepts the creation of sinks on a per job basis and allows us to wrap them with our flush-notifying implementation. 

2) How job errors are handled: Rather than polling the jobs table, we now have a Resumer implementation that wraps the main changefeed resumer and calls back into caller-provided code when the changefeed fails.

3) Since we want both of the above to happen on per-changefeed basis, we now have a `depInjector` which coordinates the above on a per-job basis.

I agree with Andrew that it might have been good to break this up into multiple commits to make it easier to review. But I also get how one change leads to the next when working in this code.

Since we are all working in the same relatively small pond at the moment, I'd err on the side of iterating on this on master.

Apologies, but while I spend a good amount of time reading this, I haven't left many in-code comments, since most of the time I found myself commenting on something it was existing code. I've marked it as approved since I'm not opposed to this direction and since it facilitated turning on more Kafka tests.
</blockquote></details>

Your overview is on point.  I would also add that #1 (flush sink), and in general the Wrap sink knob allows for much richer and more direct style of tests.  For example, instead of having "BeforeEmit" knob, you could have a "block all writes" sink.  Have the tests that need this type of functionality instantiate appropriate sink.

Regarding move commit -- I could certainly do that; it's just I heard comments the other way: why would you want to move this code when it's not clear why... 

---

**Review by stevendanna — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64693)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/64693#-M_QawOkCbL4MV9KKbGD:-M_Vfjb_AE6I-vIsv1pg:bmyh3q4) (no related file):*
>  it's just I heard comments the other way: why would you want to move this code when it's not clear why...

I usually address this by putting something like "This change will help with the changes to FOO we plan to make in subsequent commits." and if the commits are all in the same PR, then the progress is pretty clear so.

In terms of this, I'm OK with not reworking the commits as I've already read through it.  Just food for thought in the future.

---

**Comment by miretskiy** [comment]
@ajwerner Do you want to take a pass over this PR, or are you okay w/ merging this?

**Comment by miretskiy** [comment]
bors r=stevendanna
